isthmus/src/main/java/io/substrait/isthmus/type/SubstraitUserDefinedType.java

```diff
@@ -0,0 +1,251 @@
+    public static SubstraitUserDefinedAnyType from(io.substrait.type.Type.UserDefined type) {
+      return new SubstraitUserDefinedAnyType(
+          type.urn(), type.name(), type.typeParameters(), type.nullable());
+    }
+  }
+
+  /**
+   * Custom Calcite type representing a Substrait {@link
+   * io.substrait.expression.Expression.UserDefinedStruct} type.
+   *
+   * <p>This type represents a structured UDT with explicitly defined fields. Unlike {@link
+   * SubstraitUserDefinedAnyType}, the fields are accessible and can be represented as a Calcite
+   * STRUCT/ROW type with additional UDT metadata (URN, name, type parameters).
+   *
+   * <p>{@link io.substrait.expression.Expression.UserDefinedStruct UserDefinedStruct} literals use
+   * this type when passing through Calcite, preserving field structure and enabling field access.
+   * The fields are converted to Calcite literals and wrapped in a ROW type with synthetic field
+   * names (f0, f1, f2, etc.).
+   *
+   * @see SubstraitUserDefinedAnyType
+   * @see io.substrait.expression.Expression.UserDefinedStruct
+   */
+  public static class SubstraitUserDefinedStructType extends SubstraitUserDefinedType {
+
+    private final List<RelDataType> fieldTypes;
+    private final List<String> fieldNames;
+
+    public SubstraitUserDefinedStructType(
+        String urn,
+        String name,
+        List<io.substrait.proto.Type.Parameter> typeParameters,
+        boolean nullable,
+        List<RelDataType> fieldTypes,
+        List<String> fieldNames) {
+      super(urn, name, typeParameters, nullable);
+      if (fieldTypes.size() != fieldNames.size()) {
+        throw new IllegalArgumentException("Field types and names must have same length");
+      }
+      this.fieldTypes = fieldTypes;
+      this.fieldNames = fieldNames;
+    }
+
+    @Override
+    public List<RelDataTypeField> getFieldList() {
+      java.util.List<RelDataTypeField> fields = new java.util.ArrayList<>();
+      for (int i = 0; i < fieldTypes.size(); i++) {
+        fields.add(new RelDataTypeFieldImpl(fieldNames.get(i), i, fieldTypes.get(i)));
+      }
+      return fields;
+    }
+
+    @Override
+    public int getFieldCount() {
+      return fieldTypes.size();
+    }
+
+    @Override
+    public RelDataTypeField getField(String fieldName, boolean caseSensitive, boolean elideRecord) {
+      for (int i = 0; i < fieldNames.size(); i++) {
+        String name = fieldNames.get(i);
+        if (caseSensitive ? name.equals(fieldName) : name.equalsIgnoreCase(fieldName)) {
+          return new RelDataTypeFieldImpl(name, i, fieldTypes.get(i));
+        }
+      }
+      return null;
+    }
+
+    public List<RelDataType> getFieldTypes() {
+      return fieldTypes;
+    }
+
+    @Override
+    public List<String> getFieldNames() {
+      return fieldNames;
+    }
+
+    @Override
+    public SqlTypeName getSqlTypeName() {
+      // Can be considered as ROW since it has structure
+      return SqlTypeName.ROW;
```

Comment:
I am not 100% sure, but I am afraid you would then be matching built-in `ROW` types, I'd rather play it safe and use `SqlTypeName.OTHER` (like the base class) and use `isStruct()` override to indicate structure, something like:

```
@Override
public boolean isStruct() {
  return true;
}
```

I am starting from the assumption that, if any structured type would make sense as `ROW`, we would not have this boolean method, but I don't have any evidence for saying you shouldn't do it.